python/cog/server/worker.py

```diff
@@ -171,18 +178,30 @@ def run(self) -> None:
             [ws_stdout, ws_stderr], self._stream_write_hook
         )
         self._stream_redirector.start()
-
         self._setup()
-        asyncio.run(self._loop())
+        self._loop()
         self._stream_redirector.shutdown()
 
+    # I would rather have _setup_async and have a single run_until_complete for setup + loop
+    # but we need to load predictor (and handle errors) to find out the code path
     def _setup(self) -> None:
         done = Done()
         try:
             self._predictor = load_predictor_from_ref(self._predictor_ref)
+            # if the user has opted into the async world, we want to create the event loop
+            # here even if setup isn't async, it might use get_running_loop, and we want
+            # that to find the same loop that will be be used for _loop_async
+            #
+            # otherwise, the user might new_event_loop to use with a ClientSession
+            # and then try to use the same ClientSession with a different event loop
+            if is_async_predictor(self._predictor):
+                self.loop = get_loop()
             # Could be a function or a class
             if hasattr(self._predictor, "setup"):
-                run_setup(self._predictor)
+                if inspect.iscoroutinefunction(self._predictor.setup):
```

Comment:
I still think we should probably use `is_async` here?

Reply:
setup shouldn't be allowed to be a generator